servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java

```diff
@@ -203,15 +209,46 @@ public Completable apply(final int count, final Throwable t) {
                         retryWhen = retryWhen.concat(executor.timer(constant));
                     }
 
-                    return applyRetryCallbacks(retryWhen, count, t);
+                    return applyRetryCallbacksAndDraining(retryWhen, count, t);
                 }
 
                 return failed(t);
             }
 
-            Completable applyRetryCallbacks(final Completable completable, final int retryCount, final Throwable t) {
-                return retryCallbacks == null ? completable :
-                        completable.beforeOnComplete(() -> retryCallbacks.beforeRetry(retryCount, requestMetaData, t));
+            Completable applyRetryCallbacksAndDraining(final Completable completable, final int retryCount,
+                                                       final Throwable t) {
+                if (retryCallbacks == null && !(t instanceof WrappedResponseException)) {
+                    // No wrap necessary.
+                    return completable;
+                }
+                return completable.liftSync(subscriber -> new CompletableSource.Subscriber() {
+                    @Override
+                    public void onSubscribe(Cancellable cancellable) {
+                        subscriber.onSubscribe(cancellable);
+                    }
+
+                    @Override
+                    public void onComplete() {
+                        try {
+                            if (retryCallbacks != null) {
+                                retryCallbacks.beforeRetry(retryCount, requestMetaData, t);
+                            }
+                            if (t instanceof WrappedResponseException) {
+                                drainResponse(((WrappedResponseException) t).response).subscribe();
```

Comment:
We should properly drain the previous response before attempting to retry instead of subscribing and moving forward without waiting for completion. 
Imagine a use-case when users either enforce or expect a single connection. If we don't drain the previous response before retrying, it will force opening a new connection for retry and violate expectations.

Reply:
In this case, looks like we can not use `mergeDelayError` like we do in other example because we don't know the result of the returned completable from the `BackoffStrategy`. We can concat draining with that completable and add `whenCancel` operator in between to drain in case someone cancels before the retry `Completable` completes

Reply:
If our goal is to only provide the head of the message then this becomes much simpler and leak resistant. That said, will that be sufficient? I suppose it's at least a step in the right direction for users.

Reply:
I'm not sure. It's not nice to initiate a new request before we drained the previous one. Taking into account that this is not an absolute must feature (just a new feature request), no need to rush. If we don't make is sequential now, there is nothing more stable than a temporary solution 😄